docs/design/features/host-tracing.md

```diff
@@ -1,87 +1,96 @@
-Custom host can and should register the trace listener as the first thing it does with the respective host component to ensure that all tracing is routed to it.
+The functions behave exactly the same in both components. The `error_writer` parameter can be:
 
-Only one trace listener can be registered at any given time.
+* a function pointer which is then registered as the current error writer. Errors are only written to the registered error writer and will not be outputted to `stderr`.
+* `NULL` value which unregisters the previously registered error writer. After this errors go to `stderr` again.
 
-Registering custom trace listener or setting it to `NULL` doesn't override the tracing enabled by environment variables. If a trace listener is registered and the `COREHOST_TRACE=1` is set as well, the traces will be routed to both the `stderr` as well as the registered listener.
+Custom host can and should set the error writer as the first thing it does with the respective host component to ensure that all errors are routed to it.
 
-The `hostfxr` component will propagate the trace listener to the `hostpolicy` component before it calls into it. So custom host only needs to register its trace listener with the `hostfxr` component and not both. The propagation of the trace listener is only done for the duration necessary after which it will be unregistered again. So custom host might need to register its own listener if it makes calls directly to `hostpolicy` on top of the calls to `hostfxr`.
-In case of new (.NET Core 3) `hostfxr` component which would call into an old (.NET Core 2.1) `hostpolicy` component, the `hostfxr` will not perform the propagation in any way since the older `hostpolicy` doesn't support this mechanism.
+The `hostfxr_set_error_writer` sets the error writer only for the current thread (the setting is thread local). The custom host must set it separately (and potentially with different callbacks) on each thread it wants to use `hostfxr` functions (and have them redirect errors).
 
-The trace listener interface looks like this:
-``` C++
-struct host_trace_listener
-{
-    void trace_verbose(const pal::char_t *message, const pal::char_t *activityId);
-    void trace_info(const pal::char_t *message, const pal::char_t *activityId);
-    void trace_warning(const pal::char_t *message, const pal::char_t *activityId);
-    void trace_error(const pal::char_t *message, const pal::char_t *activityId);
-    void flush();
-}
+Only one error writer can be registered on a given thread at any given time.
+
+All errors are also written to a trace output if one is enabled (via `COREHOST_TRACE=1`) regardless of which error routing is used.
+
+The `hostfxr` component propagates the error writer to the `hostpolicy` component before it calls into it. So custom host only needs to register its error writer with the `hostfxr` component. The propagation of the error writer is only done for the duration necessary after which it will be unregistered again.
```

Comment:
```suggestion
The `hostfxr` component propagates the error writer to the `hostpolicy` component before it calls into it, so a custom host only needs to register its error writer with the `hostfxr` component. The propagation of the error writer is only done for the duration necessary, after which it will be unregistered again.
```